Implement redshift attenuation

TarekHC opened the pull request:
Hi Victor, 

This could be considered an extremely preliminary attempt to implement the EBL attenuation (discussed in #117 ). I also added an example macro to see the effect. I also tested to include the redshift attenuation as a free parameter, but the speed is horrible (it need to read a file with the tau values and interpolate them...).

You don't really need to merge the PR... But some feedback would be valuable.  I'm not sure how the fitting goes internally... Would this suffice?

Cheers!

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,87 @@ def __call__(self, e):
+            converters = dict()
+            for i in range(0,500):
+               converters["col%s" % i] = [ascii.convert_numpy(np.float64)]
+            taus_table = ascii.read('naima/data/tau_dominguez11.out', converters=converters)
```
If you are only reading the matrix from the file, using `np.loadtxt` might be easier than using io.ascii and building 500 converters. This is also not generalizable to other files if you want to add other models in the future.

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,87 @@ def __call__(self, e):
+            if (redshift >= 0.01):
+                taus_table['col%s' % (2+(np.abs(redshift_list-redshift)).argmin())].name = 'tau'
+                taus_table['tau'].unit = u.dimensionless_unscaled
+                taus = taus_table['tau'].quantity
```
is there a reason you can't directly do `taus = taus_table['col%s' % (2+(np.abs(redshift_list-redshift)).argmin())].quantity`?

zblz commented:
Looks good! 

A few general comments on the implementation:
- I think EBL should be in the name, as otherwise it looks like a general absorption model
- You have copied a lot of the `TableModel` logic into `AbsorptionModel`, but it might be better if you actually subclassed `TableModel` so there is no duplicate code. You can take a look at how some of the radiative models subclass `ElectronModel` and use `super` to call the `__init__` of the superclass.
- I think the table could be stored in `npz` format without headers, as it is only used internally. Renaming the column names is also probably not needed since we are not using `taus_table` anymore in the code.

Thanks for giving it a go, it will be a great addition!

zblz commented:
Parameters (redshift, model) should be stored as attributes of the class, and validated, something like:
```
self.redshift = validate_scalar(redshift, domain='positive')
# check that model has allowed value
if model not in set(['Dominguez',]):
   raise ArgumentError "Model should be one of [Dominguez, ]"
self.model = model 
```
And then use the values stored in self.

TarekHC commented:
Hi @zblz ,

I'll try to slowly implement your comments on the PR. Just one comment:

> * You have copied a lot of the TableModel logic into AbsorptionModel, but it might be better if you actually subclassed TableModel so there is no duplicate code. You can take a look at how some of the radiative models subclass ElectronModel and use super to call the __init__ of the superclass.

The only reason why I did not implement it that way is because tau values are already in log scale, so the interpolation in the y axis is in linear scale (unlike within the TableModel class, which I did not want to modify).  But we could perfectly store 10^tau values I guess...

zblz commented:
> But we could perfectly store 10^tau values I guess...

Or pass `10**(log10tau)` to `TableModel` after reading the stored `log10tau`, so you don't have to modify the table.

TarekHC commented:
Yes, I meant store in memory. Good, I will modify the PR accordingly. I'm learning a lot! :)

TarekHC commented:
Ok, I think I incorporated most of your comments.
- The table is now stored in a .npz file
- It inherits from TableModel

I also fixed an issue with the path of the file to be read. Before I was using an absolute path, and now relative to the location of the "models.py" file. Not sure if this approach is correct, but definitely better than the previous one.

One last comment: is there any way to remove the "C++-like" for loop within the "transmission" function? I'm not very familiar yet with numpy, but I'm sure there is a more efficient way of doing it.

zblz commented:
I haven't looked at it yet, but the way to load a file taking into account packaging, etc, is using:
```
from astropy.utils.data import get_pkg_data_filename
filename = get_pkg_data_filename(os.path.join('data', 'dominguez.out'))
```

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,60 @@ def __call__(self, e):
+        for i in range(0,len(data['energy'])):
+            if data['energy'][i].to('GeV').value < 1.: taus[i] = 0.
+            elif data['energy'][i].to('TeV').value > 100. : taus[i] = np.log10(6000.)
+            else: taus[i] = np.log10(super(EblAbsorptionModel,self).__call__(data['energy'][i]))
```
You don't need to use `super` here, you can just do `self.__call__`. It is needed in the init because we are overriding the init in the subclass, but since we are not overrriding `__call__`, using `self.__call__` will already point to the superclass definition.

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,60 @@ def __call__(self, e):
+        import os
+  
+        self.redshift = validate_scalar('redshift',
+                                   redshift.value,
```
Why are you using `redshift.value` rather than plain `redshift`? In the docstring you require a float for this parameter, but using `redshift.value` will fail if it is not a quantity.

TarekHC replied:
That is actually the first thing I tried, but I was getting:
"TypeError: redshift should be a scalar floating point value"
This was happening probably because I was not using any physical_type. I added the physical_type='dimensionless' to the scalar validator. 

Although now it requires a Quantity, and prompts an error when adding a float as an input... Not sure if that is what we want.

zblz replied:
Sorry, that was my mistake in suggesting `validate_scalar`. It indeed requires a quantity. You can just check that it is a scalar and positive:
```
if not np.isscalar(redshift) or redshift < 0:
   complain
```

TarekHC replied:
But that would not allow a dimensionless quantity. The ideal solution should accept both quantities and scalars, right?

The following code works for both cases, but it is probably an overkill...

if isinstance(redshift, u.Quantity):
    self.redshift = validate_scalar('redshift',
                           redshift,
                           domain='positive',
                           physical_type='dimensionless')  
elif not np.isscalar(redshift) or redshift < 0:
    raise ValueError("Redshift should be a positive scalar value.")
else: self.redshift = redshift * u.dimensionless_unscaled

cdeil commented:
@TarekHC @zblz - Status?

FYI: there's been some discussion in Gammapy concerning absorption (see e.f. https://github.com/gammapy/gammapy/issues/265#issuecomment-228359777). It would be nice to know what Naima is doing. I.e. if you can finish this PR up (or reject it ... I didn't read the discussion carefully), that would be helpful.

I think @jjlk  but maybe also @registerrier are interested to use and / or contribute to this functionality.

TarekHC commented:
Hi @cdeil , 

I implemented all @zblz suggestions, and it seems to be working. I am not sure about Victor's availability to check the code carefully, so I never insisted.

Although it is a very simple approach, maybe something fancier would be desirable... No idea.

cdeil commented:
@TarekHC - I had a quick look.

There's no test in https://github.com/zblz/naima/blob/master/naima/tests/test_models.py for `EblAbsorptionModel`. Can you add an example that executes the code and has one or a few asserts on the outputs, to make sure your code works and keeps working?

cdeil commented on naima/models.py:
```diff
@@ -431,3 +431,69 @@ def __call__(self, e):
+            redshift_list = np.arange(0.01,4,0.01)
+            energy = taus_table['energy'] * u.TeV
+            if (self.redshift >= 0.01):
+                taus = 10**taus_table['col%s' % (2+(np.abs(redshift_list-self.redshift)).argmin())] * u.dimensionless_unscaled
```
Same here, this line is very long and hard to read ... my suggestion would be to split out the `colname=` part on a separate line before.

cdeil commented on naima/models.py:
```diff
@@ -431,3 +431,69 @@ def __call__(self, e):
+    def transmission(self, data):
+        taus = np.zeros(len(data['energy']))
+        for i in range(0,len(data['energy'])):
+            if data['energy'][i].to('GeV').value < 1.: taus[i] = 0.
```
There's a few Python code style issues with your `EblAbsorptionModel ` class.

There's text editor plugins to automatically format this correctly, or command line tools to point out the issues or fix them automatically:
https://github.com/PyCQA/pycodestyle#pycodestyle-formerly-called-pep8---python-style-guide-checker
https://pypi.python.org/pypi/autopep8

(I didn't have a look at the API / implementation ... just pointing out some superfluous things.)

cdeil commented on naima/models.py:
```diff
@@ -431,3 +431,69 @@ def __call__(self, e):
+        for i in range(0,len(data['energy'])):
+            if data['energy'][i].to('GeV').value < 1.: taus[i] = 0.
+            elif data['energy'][i].to('TeV').value > 100. : taus[i] = np.log10(6000.)
+            else: taus[i] = np.log10(self.__call__(data['energy'][i]))
```
You can remove the `__call__` part, no?
I think `self.__call(something)` and `self(something)` is the same, and the second invocation is the common one.

cdeil commented on naima/models.py:
```diff
@@ -431,3 +431,69 @@ def __call__(self, e):
+            if data['energy'][i].to('GeV').value < 1.: taus[i] = 0.
+            elif data['energy'][i].to('TeV').value > 100. : taus[i] = np.log10(6000.)
+            else: taus[i] = np.log10(self.__call__(data['energy'][i]))
+        return np.exp(np.negative(taus))
```
I've never seen `np.negative` used ... it's the same as the `-` operator and that's the shorter and more common invocation, I think.

TarekHC commented:
Sure. I will do it, probably this week.

zblz commented:
Last couple of weeks have been hectic, but over this week I'll do another review, there are a couple of  API issues I have to think more about.

@cdeil, thanks for the comments!

TarekHC commented:
Well... "This week" turned into 23 days... Sorry about that!
I implemented all @cdeil  comments. I particularly liked "pycodestyle". Thank you for the tip!

I'm not sure if this test unit is enough... I just check that the transmission of redshift 0 does not modify an IC sed, and that the transmission values at z = 0.5 are smaller than the ones at z = 0. I guess at least is better than nothing!

@cdeil @zblz Let me know if you need me to improve anything else. Cheers!

zblz commented:
Travis failures are unrelated, they are because of API changes in the latest version of emcee. I am fixing those in #124, please rebase on master when that PR is merged.

zblz commented:
Can you rebase this PR on master? The emcee travis failures should be solved.

TarekHC commented:
Sorry, I was gone for the summer... 

Done!

TarekHC commented:
Ok, I also solved some errors within the commented lines (I guess for the documentation builder). Now everything should be ok.

zblz commented on examples/absorbed_SynIC.py:
```diff
@@ -0,0 +1,56 @@
+    # Initialize instances of the particle distribution and radiative models
+    BPL = BrokenPowerLaw(amplitude, 1. * u.TeV, e_break, alpha1, alpha2)
+    
+    # Compute IC on CMB and on a FIR component
```
The comment says IC on CMB and FIR but the code only does CMB.

zblz commented on examples/absorbed_SynIC.py:
```diff
@@ -0,0 +1,56 @@
+
+    # compute flux at the energies given in data['energy']
+    model = (EBL_transmitance.transmission(data) * IC.flux(data, distance=1.0 * u.kpc) + SYN.flux(data,
+                                                      distance=1.0 * u.kpc))
```
Can you break this line right after the `+` and align that with the start of the parens?

zblz commented on examples/absorbed_SynIC.py:
```diff
@@ -0,0 +1,56 @@
+
+# Define the redshift of the source, and absorption model
+redshift = 0.06 * u.dimensionless_unscaled
+EBL_transmitance = EblAbsorptionModel(redshift, 'Dominguez')
```
Can you move the definition of these inside the function as with the other models defined? That makes the function more portable if somebody wants to reuse it.

zblz commented on examples/absorbed_SynIC.py:
```diff
@@ -0,0 +1,56 @@
+redshift = 0.06 * u.dimensionless_unscaled
+EBL_transmitance = EblAbsorptionModel(redshift, 'Dominguez')
+
+def ElectronSynIC(pars, data):
```
The name should include `absorbed` or `EBL`

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,76 @@ def __call__(self, e):
+    ----------
+    redshift : float
+        Redshift considered for the absorption evaluation.
+    ebl_absorption_model : string
```
add `, optional` after `string`

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,76 @@ def __call__(self, e):
+    -----
+    Dominguez model refers to the Dominguez 2011 EBL model. Current implementation 
+    does NOT perform an interpolation in the redshift, so it just uses the closest
+    z value from the finely binned tau_dominguez11.out file (delta_z=0.01).
```
the filename is now `npz`, probably should change that here so somebody looking for the file in the source can find it.

zblz commented on naima/tests/test_models.py:
```diff
@@ -4,37 +4,38 @@
 
 from ..utils import trapz_loglog
-
+ 
```
You have added a space on *all* empty lines on this file. Can you remove them?

zblz commented on naima/tests/test_models.py:
```diff
@@ -523,3 +524,32 @@ def test_tablemodel():
+    from ..radiative import InverseCompton
+    IC = InverseCompton(BPL,
+                        seed_photon_fields=['CMB'],
+                        Eemin=10 * u.GeV)
```
Here we don't need to test the absorption on a computed IC spectrum because we only need to test the absorption, `InverseCompton` and `BrokenPowerLaw` are tested elsewhere. Can you use `np.ones_like(e)` as your source spectrum?

zblz commented on naima/tests/test_models.py:
```diff
@@ -523,3 +524,32 @@ def test_tablemodel():
+    assert_allclose(IC.sed(e).value, IC.sed(e).value * EBL_zero.transmission(e), rtol=1e-1)
+#   Make sure the transmission at z = 0. is always larger than the one at z = 0.5
+    difference = EBL_zero.transmission(e) - EBL_moderate.transmission(e)
+    assert(not np.any(difference < -1E-10))
```
`not np.any` can be replaced by `np.all`.

zblz commented on naima/tests/test_models.py:
```diff
@@ -4,37 +4,38 @@
 from numpy.testing import assert_allclose
 from astropy.tests.helper import pytest
 from astropy.extern import six
+from sre_constants import ASSERT_NOT
```
`ASSERT_NOT` is not being used in this file, remove this import.

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,76 @@ def __call__(self, e):
+            redshift_list = np.arange(0.01, 4, 0.01)
+            energy = taus_table['energy'] * u.TeV
+            if (self.redshift >= 0.01):
+                colname = 'col%s' % (2+(np.abs(redshift_list-self.redshift)).argmin())
```
Please add spaces around all binary operators.

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,76 @@ def __call__(self, e):
+        if self.model == 'Dominguez':
+            filename = get_pkg_data_filename(os.path.join('data', 'tau_dominguez11.npz'))
+            taus_table = np.load(filename)['arr_0']
+            redshift_list = np.arange(0.01, 4, 0.01)
```
Where is this redshift list extracted from? It would be good to have a comment here about the structure of `tau_dominguez11`.

zblz commented on naima/models.py:
```diff
@@ -431,3 +431,76 @@ def __call__(self, e):
+        elif not np.isscalar(redshift) or redshift < 0:
+            raise ValueError("Redshift should be a positive scalar value.")
+        else:
+            self.redshift = redshift * u.dimensionless_unscaled
```
You should run the quantity through `validate_scalar` after adding the `dimensonless_unscaled` units.
Maybe something like (pseudocode):
```
if not isinstance(redshift, u.Quantity):
    redshift *= u.dimensionless_unscaled

self.redshift = validate_scalar(...)
```
(validate scalar will check that it is scalar and positive, so no need to do it separately)

zblz commented on examples/absorbed_SynIC.py:
```diff
@@ -0,0 +1,56 @@
+from naima.models import InverseCompton, Synchrotron, ExponentialCutoffPowerLaw, BrokenPowerLaw, EblAbsorptionModel
+
+# Define the redshift of the source, and absorption model
+redshift = 0.06 * u.dimensionless_unscaled
```
The class can take a float now, so change this to `redshift = 0.06`

zblz commented:
Thanks for the update, it is looking good! I have added some inline comments.

TarekHC commented:
Sorry to take a month for these changes... I was terribly busy!

I believe I fixed all issues. Let me know if you have more comments!

TarekHC commented:
Hi @zblz , 

Is there anything else missing within this PR?

Cheers!

zblz commented:
Last changes are looking good and the fail on Travis was bogus, restarting it took it away. Merging now.

Thanks for the contribution!

TarekHC commented:
Awesome, thanks!
